File: LinuxRT/Clocked_Rate/wiringPi_clocked.c
//  Run at desired frequencies
//  Program in C, using WiringPi library from Gordon Henderson (now deprecated)

//  Based on blink.c from Gordon Henderson
//      see here : http://wiringpi.com/examples/blink/
//  I use WiringPi pin numbering scheme. 
//  Also here, we investigate the capability to reach some desired frequencies with Delays or Timings based algorithms


#include <stdio.h>
#include <time.h>
#include <unistd.h>
#include <string.h>
#include <wiringPi.h>

#define PIN 24 /// unused here
#define POUT 7


int main ()
{
		if (wiringPiSetup () == -1) // init WP
			return 1 ;

		pinMode(POUT, OUTPUT); // aka BCM_GPIO pin 17	


		// INITIALISATIONS
		short val = 0; // value of pin to set 
		// Note : these periods are half of the true period !!!
		long long period_us = 500000; // in us, period to set
		long long period_ns = 500000*1000; // in ns, period to set
		long long startTime_s; // Used to compare timings for the duration of the test
		int timeTest = 5; // last 5s
		unsigned int pause = 1*1000000; // last 1s 
		struct timespec sysTime; // used to get TimeOfDay exact at the nanosecond for the duration of periods
		struct timespec sleepTime, remTime; // used to set nanosleep() function ===> http://man7.org/linux/man-pages/man2/nanosleep.2.html , Here we can get the "error" through remTime ?
			sleepTime.tv_sec = 0; // WARNING : here we dont plan to used a period over the second, so only care about 'tv_nsec'.    If you want a period above second, use 'usleep(period_us)'.
			sleepTime.tv_nsec = (long)period_ns; 
		long long refTime_s, refTime_ns; // Used to compare timings for the duration of periods
		long long list_periods_ns[] = {500000000, 250000000, 100000000, 50000000, 25000000, 10000000, 5000000, 2500000, 1000000, 500000, 250000, 100000, 50000, 25000, 10000, 5000, 2500, 1000, 500, 250}; // all these periods to bench
		long long count; // used to count the real number of periods achieved during the test
		double res[2][20]; // array of results depending the method


		printf("===> Starting...\n\n"); 

		for(int i = 0; i<20; i++)
		{
				period_ns = list_periods_ns[i]; // every loop a new period is set for benchmarking the method
				sleepTime.tv_nsec = (long)period_ns; 
				if(i<18)
					period_us = list_periods_ns[i]/1000; // 500ns and 250ns can't be achieved with microsecond resolution (only Method 1 can (if you want) use microsecond)


				printf("---> Method 1 : with delays (resolution in ns)\n");

// Method 1 : DELAYs
//	---------------------------------------------------------------
//	---------- Others process step in during execution : ----------
//	--------------- During "usleep", the SCHEDULER pass work ------
//	--------------- on others process of the system !!! -----------
//	-------- Into the "do while" loop, instructions take time -----
//	--- Here we use fwrite, a buffered function of stdio (in C) ---
//	---------------------------------------------------------------	
				count=0;
				clock_gettime(CLOCK_REALTIME, &sysTime);
				startTime_s = sysTime.tv_sec;
				while(sysTime.tv_sec < (startTime_s+timeTest)){
					clock_gettime(CLOCK_REALTIME, &sysTime);
					
					val = 1 - val; // algo balance
					digitalWrite(POUT, val);
					//usleep(period_us);
					nanosleep(&sleepTime, &remTime);
					count++;
				}

				//Save result
				res[0][i] = (double)(count)/5/2;
//	---------------------------------------------------------------


				// Pause
				printf("pause of %d second...\n", pause/1000000);
				usleep(pause);


				printf("---> Method 2 : with timing (resolution=ns)\n");
					
// Method 2 : TIMING
//	---------------------------------------------------------------
//	---------- No others process step in during execution ---------
//	-------- Into the "if" condition, instructions take time ------
//	--- Here we use fwrite, a buffered function of stdio (in C) ---
//	---------------------------------------------------------------
				count=0;
				clock_gettime(CLOCK_REALTIME, &sysTime);
				refTime_ns = sysTime.tv_nsec;
				startTime_s = sysTime.tv_sec;
				refTime_s = startTime_s;
				while(sysTime.tv_sec<(startTime_s+timeTest)){
					clock_gettime(CLOCK_REALTIME, &sysTime);
					
					if(((sysTime.tv_sec == refTime_s) && (sysTime.tv_nsec >= (refTime_ns + period_ns))) || (sysTime.tv_sec > refTime_s)) 
					{
						refTime_ns = sysTime.tv_nsec;
						refTime_s = sysTime.tv_sec;
						val = 1 - val; // algo balance
						digitalWrite(POUT, val);
						count++;
					}
				}
				
				//Save result
				res[1][i] = (double)count/5/2;
// 	---------------------------------------------------------------


				// Pause
				printf("pause of %d second...\n", pause/1000000);
				usleep(pause);

		}


// 	---------------------------------------------------------------
//	--------- This part display the results : ---------------------
//	---------     -> row = frequency adjusted ---------------------
//	---------     -> column = Method used -------------------------
// 	---------------------------------------------------------------
		printf("------- BENCHMARK's RESULTS -------\n");
		printf(" \t\t\t Method 1 \t\t\t Method 2\n");

		for(int i=0; i<20; i++)
		{
			for(int j=0; j<2; j++)
			{
				printf(" \t\t\t %.3f", res[j][i]);
			}
			printf("\n\n");
		}
// 	---------------------------------------------------------------


		printf("finished !\n");

		return(0);

}

File: LinuxRT/Clocked_Rate/wiringPi_clocked.py
#!/usr/bin/python3
#-*- coding: utf-8 -*-


#   Run at desired frequencies
#   Program in Python, using unofficial port of Gordon's WiringPi library
#       see here : https://github.com/WiringPi/WiringPi-Python
#   I use WiringPi pin numbering scheme. 
#   Also here, we investigate the capability to reach some desired frequencies with Delays or Timings based algorithms






import numpy
import wiringpi
import time

pin = 7
val = 1
pause_s = 1 # Pause of 1s between tests
test_Time = 5.0 # test will lasts 5s
period_s = 0.0005 # Actually half of the true period
period_ns = 500*1000 # half of the true period
list_periods = [0.5, 0.25, 0.1, 0.05, 0.025, 0.01, 0.005, 0.0025, 0.001, 0.0005, 0.00025, 0.0001, 0.00005, 0.000025, 0.00001, 0.000005, 0.0000025, 0.000001, 0.0000005, 0.00000025]
start_Time_s = ref_Time_s = 0.0 # in second
ref_Time_ns = 0 # in nanosecond
count = 0 # count / 5 = Hz
res = numpy.zeros((20,2), dtype='f')


# Initialisation de wiringPi... 
wiringpi.wiringPiSetup()

# Activate the pin
wiringpi.pinMode(pin, 1)



print('==========____START____==========')


# Auto loop on diff period...
for i in range(20):
    period_s =list_periods[i]
    period_ns = list_periods[i]*1000000000


    print('Method 1 : Delays... use double in seconds')
    # ----------------------------------------
    # First method : Delays
    # ----------------------------------------
    count = 0
    start_Time_s = time.perf_counter()
    while time.perf_counter() < (start_Time_s+test_Time):
        val = 1 - val # algo balance
        wiringpi.digitalWrite(pin,val)
        count = count + 1
        time.sleep(period_s)
    print('Method 1 = {var1:.6f} Hz / {var2:.0f} Hz. ==> {var3:.6f} %'.format(var1= count/5/2, var2 = 1/2/period_s, var3 = count/5/2*2*period_s*100))
    res [i][0] = count/5/2
    # ----------------------------------------



    # Pause = 1s
    print('Pause of 1s')
    time.sleep(pause_s)


    print('Method 2 : Timings... use double in seconds')
    # ----------------------------------------:
    # Second method : Timings (s)
    # ----------------------------------------
    count = 0
    start_Time_s = ref_Time_s = time.perf_counter()
    while time.perf_counter() < (start_Time_s+test_Time):
        if  (ref_Time_s+period_s) < time.perf_counter():
            ref_Time_s = time.perf_counter()
            val = 1 - val # algo balance
            wiringpi.digitalWrite(pin,val)
            count = count + 1
    print('Method 2 = {var1:.6f} Hz / {var2:.0f} Hz. ==> {var3:.6f} %'.format(var1= count/5/2, var2 = 1/2/period_s, var3 = count/5/2*2*period_s*100))
    res [i][1] = count/5/2
    # ----------------------------------------


    # Pause = 1s
    print('Pause of 1s')
    time.sleep(pause_s)

# ----------------------------------------
# --------- PRINTING OF RESULTS ----------
# ----------------------------------------
print('\n\t------- Benchmark\'s results -------\n')
print('\t\t\t Method 1\t\t\t | Method 2\n\n')

for i in range(20):
    print('\t\t\t {var1:.3f} Hz\t\t\t | {var2:.3f} Hz\n\n'.format(var1 = res[i][0], var2 = res[i][1]))
# ----------------------------------------


print('==========____FINISHED____==========')


